### DISCUSSION TITLE
Delete Plugin Tables Function Help

### DISCUSSION THREAD

**Discussion by aljawaid** [aljawaid/ContentCleaner#13] [Q&A] [answered]
@alfredbuehler @creecros could you please help?

All the frontend is done, a button will be created to delete all the plugin tables for the plugin pulled from the JSON.

### Controller:
https://github.com/aljawaid/ContentCleaner/blob/ea9d7e22fbc6c1fa32e59e73a6f93e4d010f88ec/Controller/PluginCleaningController.php#L48-L74

### Model is here:
https://github.com/aljawaid/ContentCleaner/blob/ea9d7e22fbc6c1fa32e59e73a6f93e4d010f88ec/Model/ApplicationCleaningModel.php#L254-L271

### Button:
https://github.com/aljawaid/ContentCleaner/blob/ea9d7e22fbc6c1fa32e59e73a6f93e4d010f88ec/Template/config/modals/plugin_deep_clean.php#L68

**Would this work?**

**Comment by alfredbuehler** [comment]
Regarding the Model: Think about adding a dry-run mode, that just logs the SQL statements to execute. I'm not very familiar with all SQL dialects.

**Comment by alfredbuehler** [comment]
There's something wrong: After installing Wiki:
![image](https://github.com/aljawaid/ContentCleaner/assets/2079289/ca0a5a9e-183b-4d55-898e-eafc73fd6914)
These tables cannot be associated to Wiki, but they should.
As a consequence, **these tables can be deleted, even if Wiki is installed!!!**

**Reply by aljawaid** [reply]
Yes, once the plugin is uninstalled, the tables are not recognised. According to @creecros there is no way to link them, so I put the notice for users here:
> Jobs in this section affect leftover tables which were created by plugins. Once a plugin is uninstalled, all leftover tables for the uninstalled plugin are marked as unknown. Deleting plugin related database tables before uninstalling the plugin will prompt the plugin to recreate the tables, not uninstall the plugin.

**Reply by creecros** [reply]
If wiki is installed they should not show as "unknown". If thats the case, something is wrong.

**Reply by aljawaid** [reply]
> If wiki is installed they should not show as "unknown". If thats the case, something is wrong.

oh hello :)
I understood Alfred's comment as he had uninstalled Wiki... maybe I was mistaken.

Anyhow, none of them features have been changed and it looks okay on my side... the only thing which is messed up/work in progress is the plugin cleaning controller and functions.

**Reply by creecros** [reply]
Ya, I've been kind of busy, but one day I'll have time again, once I settle in.

**Comment by alfredbuehler** [comment, marked as answer]
After analyzing things, my suggestion for the delete function looks like this:
```php
public function deletePluginTables()
{
    $this->checkCSRFParam();

    // Pull variable from button
    $plugin_job_name = $this->request->getStringParam('plugin_job_name');
    $plugin_tables = [];

    // Match variable to json content
    foreach ($this->helper->pluginCleaningHelper->getDeletablePlugins() as $plugin) {
        if ($plugin['plugin_title'] === $plugin_job_name && isset($plugin['plugin_tables'])) {
            $plugin_tables = $plugin['plugin_tables'];
            break;
        }
    }

    // Delete the tables
    $success = true;
    foreach ($plugin_tables as $table) {
        if (!$this->applicationCleaningModel->delete($table)) {
            $success = false;
            break;
        }
    }

    if ($success) {
        $this->flash->success(t('DEEP CLEANING COMPLETE: Plugin tables were deleted successfully'));
    } else {
        $this->flash->failure(t('DEEP CLEANING FAILED: Plugin tables could not be deleted'));
    }

    $this->response->redirect($this->helper->url->to('ContentCleanerController', 'show', array('plugin' => 'ContentCleaner')));
}
```
Unfortunately, I cannot do more. Sorry.

**Reply by aljawaid** [reply]
that's great! so to understand it...

The button will provide the plugin name, match that to the foreach to get the plugin tables for that particular plugin. Then for each plugin table found, use the applicationCleaningModel to delete each table

**Reply by aljawaid** [reply]
Commit https://github.com/aljawaid/ContentCleaner/commit/76545b07a630b368b373d85c45dba43fb4b8ef24

**Reply by aljawaid** [reply]
@alfredbuehler This codeblock provides a success message even when the relevant plugin tables do not exist.

I tried to change it like this, but now all I get is a false message.  Any insight?
```php
/**
     * Delete Plugin Tables
     *
     * @param   $plugin_job_name                string
     * @param   $plugin_tables                  array
     * @author  alfredbuehler Alfred Bühler
     */
    public function deletePluginTables()
    {
        $this->checkCSRFParam();

        // Pull the variable from the button
        $plugin_job_name = $this->request->getStringParam('plugin_job_name');
        $plugin_tables = [];

        // Match the variable to the JSON content if the plugin name matches
        foreach ($this->helper->pluginCleaningHelper->getDeletablePlugins() as $plugin) {
            if ($plugin['plugin_title'] === $plugin_job_name && isset($plugin['plugin_tables'])) {
                $plugin_tables = $plugin['plugin_tables'];
                break;
            }
        }

        $existing_tables = $this->applicationCleaningModel->getTables();

        // Delete the plugin tables
        $success = true;
        foreach ($plugin_tables as $table) {
            if (!in_array($table, $existing_tables)) {
                $success = false;
                break;
            } elseif (!$this->applicationCleaningModel->delete($table)) {
                $success = false;
                break;
            } elseif ($this->applicationCleaningModel->delete($table)) {
                $success = true;
            }
        }

        if ($success) {
            $this->flash->success(t('DEEP CLEANING COMPLETE: Plugin tables were deleted successfully for ' . $plugin_job_name . ''));
        } else {
            $this->flash->failure(t('DEEP CLEANING FAILED: Plugin tables could not be deleted for ' . $plugin_job_name . ''));
        }

        $this->response->redirect($this->helper->url->to('ContentCleanerController', 'show', array('plugin' => 'ContentCleaner')));
    }
```

**Reply by aljawaid** [reply]
#14

**Reply by aljawaid** [reply]
Fixed through helper https://github.com/aljawaid/ContentCleaner/commit/ff56637d16a9e3ab090ce3c4a0b006836a482cf3

Button is now disabled if no tables are found
